gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/contract/AbstractGoogleContractConcatTest.java

```diff
@@ -0,0 +1,36 @@
+package com.google.cloud.hadoop.fs.gcs.contract;
+
+import com.google.cloud.hadoop.fs.gcs.GoogleHadoopFileSystemBase;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.contract.AbstractContractConcatTest;
+import org.apache.hadoop.fs.contract.ContractTestUtils;
+import org.junit.Test;
+
+import static org.apache.hadoop.fs.contract.ContractTestUtils.assertFileHasLength;
+import static org.apache.hadoop.fs.contract.ContractTestUtils.createFile;
+import static org.apache.hadoop.fs.contract.ContractTestUtils.dataset;
+
+public abstract class AbstractGoogleContractConcatTest extends AbstractContractConcatTest {
+  @Test
+  public void testConcatMultiple() throws Throwable {
+    int numFiles = GoogleHadoopFileSystemBase.MAX_COMPOSE_OBJECTS * 3 / 2;
+    Path testPath = path("test");
+
+    byte[][] blocks = new byte[numFiles][0];
+    Path[] srcs = new Path[numFiles];
+    for (int i = 0; i < numFiles; i++) {
+      Path srcFile = new Path(testPath, "" + i);
+      blocks[i] = dataset(TEST_FILE_LEN, i, 255);
+      createFile(getFileSystem(), srcFile, true, blocks[i]);
+      srcs[i] = srcFile;
+    }
+    Path target = new Path(testPath, "target");
+
+    createFile(getFileSystem(), target, false, new byte[0]);
```

Comment:
For my understanding, does that mean Hadoop api requires target path to exists before call to concat?

Reply:
Yes it does. The [javadoc](https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-common/api/org/apache/hadoop/fs/FileSystem.html#concat(org.apache.hadoop.fs.Path,%20org.apache.hadoop.fs.Path[])) is not very clear on this point, but the HDFS implementation certainly requires that the target exists. There is also a filesystem contract test for exactly this case (`AbstractContractConcatTest#testConcatMissingTarget`).